File: azdo-pbi-analyzer/src/analyzers/requirements-analyzer.ts
import { PBIData } from '../models/pbi-data.js';

/**
 * Analyzes PBIs to extract structured requirements and deliverables
 */
export class RequirementsAnalyzer {
    /**
     * Extract structured requirements from PBI description
     */
    public extractRequirements(pbi: PBIData): string[] {
        const requirements: string[] = [];
        const description = this.stripHtml(pbi.description);

        // Look for numbered lists or bullet points
        const listPatterns = [
            /\d+\.\s+(.+?)(?=\n\d+\.|\n-|\n\*|$)/g,
            /[-*]\s+(.+?)(?=\n-|\n\*|\n\d+\.|$)/g,
        ];

        for (const pattern of listPatterns) {
            const matches = description.matchAll(pattern);
            for (const match of matches) {
                const req = match[1].trim();
                if (req.length > 10 && !requirements.includes(req)) {
                    requirements.push(req);
                }
            }
        }

        // If no structured list found, extract sentences with requirement keywords
        if (requirements.length === 0) {
            const requirementKeywords = [
                'must', 'should', 'shall', 'need to', 'required to',
                'enable', 'allow', 'support', 'provide', 'implement'
            ];

            const sentences = description.split(/[.!?]+/);
            for (const sentence of sentences) {
                const lowerSentence = sentence.toLowerCase();
                if (requirementKeywords.some(kw => lowerSentence.includes(kw))) {
                    const trimmed = sentence.trim();
                    if (trimmed.length > 20) {
                        requirements.push(trimmed);
                    }
                }
            }
        }

        return requirements.slice(0, 10); // Limit to top 10 requirements
    }

    /**
     * Extract deliverables from acceptance criteria
     */
    public extractDeliverables(pbi: PBIData): string[] {
        const deliverables: string[] = [];
        const criteria = this.stripHtml(pbi.acceptanceCriteria);

        if (!criteria) {
            return ['Implementation of all requirements', 'Unit tests', 'Documentation'];
        }

        // Extract checkbox items or numbered lists
        const patterns = [
            /\[[ x]\]\s+(.+?)(?=\n|$)/gi,
            /\d+\.\s+(.+?)(?=\n\d+\.|\n-|$)/g,
            /[-*]\s+(.+?)(?=\n-|\n\*|$)/g,
        ];

        for (const pattern of patterns) {
            const matches = criteria.matchAll(pattern);
            for (const match of matches) {
                const deliverable = match[1].trim();
                if (deliverable.length > 5) {
                    deliverables.push(deliverable);
                }
            }
        }

        return deliverables.length > 0
            ? deliverables
            : this.extractRequirements(pbi);
    }

    /**
     * Extract dependencies from description and related items
     */
    public extractDependencies(pbi: PBIData): string[] {
        const dependencies: string[] = [];

        // Check related items
        const parentItems = pbi.relatedItems.filter(r => r.relationship === 'Parent');
        const predecessors = pbi.relatedItems.filter(r => r.relationship === 'Predecessor');

        if (parentItems.length > 0) {
            dependencies.push(...parentItems.map(p => `Parent: ${p.title} (${p.type})`));
        }

        if (predecessors.length > 0) {
            dependencies.push(...predecessors.map(p => `Depends on: ${p.title} (${p.type})`));
        }

        // Extract from description
        const description = this.stripHtml(pbi.description).toLowerCase();
        const dependencyKeywords = [
            'depends on', 'requires', 'prerequisite', 'blocked by', 'needs'
        ];

        for (const keyword of dependencyKeywords) {
            if (description.includes(keyword)) {
                const sentences = pbi.description.split(/[.!?]+/);
                for (const sentence of sentences) {
                    if (sentence.toLowerCase().includes(keyword)) {
                        dependencies.push(this.stripHtml(sentence).trim());
                        break;
                    }
                }
            }
        }

        return dependencies.length > 0
            ? dependencies
            : ['No explicit dependencies identified'];
    }

    /**
     * Determine functional area from tags and area path
     */
    public getFunctionalArea(pbi: PBIData): string {
        // Check tags first
        if (pbi.tags.length > 0) {
            return pbi.tags[0];
        }

        // Extract from area path
        const areaPathParts = pbi.areaPath.split('\\');
        return areaPathParts[areaPathParts.length - 1] || 'General';
    }

    /**
     * Determine if PBI is high priority/critical
     */
    public isCritical(pbi: PBIData): boolean {
        return pbi.priority === 1 ||
            pbi.tags.some(t => ['critical', 'high-priority', 'p0', 'p1'].includes(t.toLowerCase()));
    }

    /**
     * Strip HTML tags from text
     */
    private stripHtml(html: string): string {
        return html
            .replace(/<[^>]*>/g, ' ')
            .replace(/&nbsp;/g, ' ')
            .replace(/&amp;/g, '&')
            .replace(/&lt;/g, '<')
            .replace(/&gt;/g, '>')
            .replace(/\s+/g, ' ')
            .trim();
    }

    /**
     * Get business value summary
     */
    public getBusinessValue(pbi: PBIData): string {
        if (pbi.businessValue) {
            return `Business Value: ${pbi.businessValue}`;
        }

        // Try to extract from description
        const description = this.stripHtml(pbi.description).toLowerCase();
        const valueKeywords = ['business value', 'impact', 'benefit', 'roi'];

        for (const keyword of valueKeywords) {
            if (description.includes(keyword)) {
                const sentences = pbi.description.split(/[.!?]+/);
                for (const sentence of sentences) {
                    if (sentence.toLowerCase().includes(keyword)) {
                        return this.stripHtml(sentence).trim();
                    }
                }
            }
        }

        return 'Business value not explicitly defined';
    }
}

File: azdo-pbi-analyzer/src/AzDoPbiAnalyzer.Core/Analyzers/RequirementsAnalyzer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using AzDoPbiAnalyzer.Core.Models;

namespace AzDoPbiAnalyzer.Core.Analyzers;

public class RequirementsAnalyzer
{
    public List<string> ExtractRequirements(PBIData pbi)
    {
        var requirements = new List<string>();
        var description = StripHtml(pbi.Description);

        // Look for numbered lists or bullet points
        var listPatterns = new[]
        {
            @"\d+\.\s+(.+?)(?=\n\d+\.|\n-|\n\*|$)",
            @"[-*]\s+(.+?)(?=\n-|\n\*|\n\d+\.|$)"
        };

        foreach (var pattern in listPatterns)
        {
            var matches = Regex.Matches(description, pattern, RegexOptions.Singleline);
            foreach (Match match in matches)
            {
                var req = match.Groups[1].Value.Trim();
                if (req.Length > 10 && !requirements.Contains(req))
                {
                    requirements.Add(req);
                }
            }
        }

        // If no structured list found, extract sentences with keywords
        if (requirements.Count == 0)
        {
            var keywords = new[]
            {
                "must", "should", "shall", "need to", "required to",
                "enable", "allow", "support", "provide", "implement"
            };

            var sentences = description.Split(new[] { '.', '!', '?' }, StringSplitOptions.RemoveEmptyEntries);
            foreach (var sentence in sentences)
            {
                var lower = sentence.ToLower();
                if (keywords.Any(k => lower.Contains(k)))
                {
                    var trimmed = sentence.Trim();
                    if (trimmed.Length > 20)
                    {
                        requirements.Add(trimmed);
                    }
                }
            }
        }

        return requirements.Take(10).ToList();
    }

    public List<string> ExtractDeliverables(PBIData pbi)
    {
        var deliverables = new List<string>();
        var criteria = StripHtml(pbi.AcceptanceCriteria);

        if (string.IsNullOrWhiteSpace(criteria))
        {
            return new List<string> { "Implementation of all requirements", "Unit tests", "Documentation" };
        }

        var patterns = new[]
        {
            @"\[[ x]\]\s+(.+?)(?=\n|$)",
            @"\d+\.\s+(.+?)(?=\n\d+\.|\n-|$)",
            @"[-*]\s+(.+?)(?=\n-|\n\*|$)"
        };

        foreach (var pattern in patterns)
        {
            var matches = Regex.Matches(criteria, pattern, RegexOptions.Singleline | RegexOptions.IgnoreCase);
            foreach (Match match in matches)
            {
                var deliverable = match.Groups[1].Value.Trim();
                if (deliverable.Length > 5)
                {
                    deliverables.Add(deliverable);
                }
            }
        }

        return deliverables.Count > 0 ? deliverables : ExtractRequirements(pbi);
    }

    public List<string> ExtractDependencies(PBIData pbi)
    {
        var dependencies = new List<string>();

        // Related items
        var parents = pbi.RelatedItems.Where(r => r.Relationship == "Parent");
        var predecessors = pbi.RelatedItems.Where(r => r.Relationship == "Predecessor");

        dependencies.AddRange(parents.Select(p => $"Parent: {p.Title} ({p.Type})"));
        dependencies.AddRange(predecessors.Select(p => $"Depends on: {p.Title} ({p.Type})"));

        // Description analysis
        var description = StripHtml(pbi.Description).ToLower();
        var keywords = new[] { "depends on", "requires", "prerequisite", "blocked by", "needs" };

        foreach (var keyword in keywords)
        {
            if (description.Contains(keyword))
            {
                var sentences = pbi.Description.Split(new[] { '.', '!', '?' }, StringSplitOptions.RemoveEmptyEntries);
                foreach (var sentence in sentences)
                {
                    if (sentence.ToLower().Contains(keyword))
                    {
                        dependencies.Add(StripHtml(sentence).Trim());
                        break;
                    }
                }
            }
        }

        return dependencies.Count > 0 ? dependencies : new List<string> { "No explicit dependencies identified" };
    }

    public bool IsCritical(PBIData pbi)
    {
        return pbi.Priority == 1 ||
               pbi.Tags.Any(t => new[] { "critical", "high-priority", "p0", "p1" }.Contains(t.ToLower()));
    }

    private string StripHtml(string html)
    {
        if (string.IsNullOrEmpty(html)) return string.Empty;
        
        var text = Regex.Replace(html, "<[^>]*>", " ");
        text = text.Replace("&nbsp;", " ")
                   .Replace("&amp;", "&")
                   .Replace("&lt;", "<")
                   .Replace("&gt;", ">")
                   .Replace("  ", " ");
                   
        return text.Trim();
    }
}
